src/agent/coverage/src/block/windows.rs

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use std::collections::BTreeMap;
+use std::process::Command;
+use std::time::{Duration, Instant};
+
+use anyhow::Result;
+use debugger::{
+    dbghelp::SymInfo,
+    debugger::{BreakpointId, BreakpointType, DebugEventHandler, Debugger},
+    target::Module,
+};
+
+use crate::{AppCoverageBlocks, ModuleCoverageBlocks};
+
+pub fn record(cmd: Command) -> Result<AppCoverageBlocks> {
+    let mut handler = BlockCoverageHandler::new();
+
+    let (mut dbg, _child) = Debugger::init(cmd, &mut handler)?;
+    dbg.run(&mut handler)?;
+
+    Ok(handler.coverage)
+}
+
+pub struct BlockCoverageHandler {
+    bp_to_block: BTreeMap<BreakpointId, (usize, usize)>,
+    coverage: AppCoverageBlocks,
+    started: Instant,
+    max_duration: Duration,
+    timed_out: bool,
+}
+
+impl BlockCoverageHandler {
+    pub fn new() -> Self {
+        let coverage = AppCoverageBlocks::new();
+        let bp_to_block = BTreeMap::default();
+        let started = Instant::now();
+        let max_duration = Duration::from_secs(5);
+        let timed_out = false;
+
+        Self {
+            bp_to_block,
+            coverage,
+            max_duration,
+            started,
+            timed_out,
+        }
+    }
+
+    pub fn pc(&self, dbg: &mut Debugger) -> Result<(u64, Option<SymInfo>)> {
+        let rip = iced_x86::Register::RIP;
+        let pc = dbg.read_register_u64(rip)?;
+        let sym = dbg.get_symbol(pc).ok();
+
+        Ok((pc, sym))
+    }
+
+    fn add_module(&mut self, dbg: &mut Debugger, module: &Module) {
+        let bitset = crate::pe::process_image(module.path(), false);
+
+        // If we can't add the module, continue debugging.
+        // We don't expect to have symbols for every module.
+        if let Err(err) = bitset {
+            log::warn!(
+                "cannot record coverage for module = {}, err = {}",
+                module.path().display(),
+                err,
+            );
+            return;
+        }
+
+        // Won't panic, due to above check.
+        let bitset = bitset.unwrap();
```

Comment:
Maybe use `and_then`/`or_else` or`match` instead of `unwrap?

Reply:
Here's the `match` version:
```rust
let bitset = match bitset {
    Ok(bitset) => bitset,
    Err(err) => {
        log::warn!(
            "cannot record coverage for module = {}, err = {}",
            module.path().display(),
            err,
        );
        return;                
    }
};
```
I think it's often a bit more surprising to see an early return when using a `match` _expression_, vs. statement.

In the `if let`, the early return pattern felt more blatant. But maybe it's still better to not have an unreachable `unwrap()` to justify (and whose unreachability we'd need to preserve).

Which do you prefer?

Reply:
I honestly prefer to read the `if let` version, but from a future maintainability point of view, prefer the match.